stdlib/grpc/src/main/java/org/ballerinalang/net/grpc/listener/StreamingServerCallHandler.java

```diff
@@ -43,57 +58,87 @@
-        Map<String, Object> properties = new HashMap<>();
-        if (ObserveUtils.isObservabilityEnabled()) {
-            properties.put(ObservabilityConstants.KEY_OBSERVER_CONTEXT, context);
+    private StreamObserver invoke(StreamObserver responseObserver, ServerCall call) {
+        ObserverContext context = call.getObserverContext();
+        ObjectValue streamIterator = BallerinaValues.createObjectValue(GrpcConstants.PROTOCOL_GRPC_PKG_ID,
+                ITERATOR_OBJECT_NAME, new Object[1]);
+        Semaphore listenerSemaphore = new Semaphore(1, true);
+        Semaphore iteratorSemaphore = new Semaphore(0, true);
+        streamIterator.addNativeData(LISTENER_LOCK, listenerSemaphore);
+        streamIterator.addNativeData(ITERATOR_LOCK, iteratorSemaphore);
+        streamIterator.addNativeData(CLIENT_ENDPOINT_TYPE, getConnectionParameter(responseObserver));
+        StreamValue requestStream = new StreamValue(new BStreamType(inputType), streamIterator);
+        onStreamInvoke(resource, requestStream, call.getHeaders(), responseObserver, context);
+        return new StreamingServerRequestObserver(streamIterator, iteratorSemaphore, listenerSemaphore);
+    }
+
+    private static final class StreamingServerRequestObserver implements StreamObserver {
+        
+        private final ObjectValue streamIterator;
+        private final Semaphore iteratorSemaphore;
+        private final Semaphore listenerSemaphore;
+
+        StreamingServerRequestObserver(ObjectValue streamIterator, Semaphore iteratorSemaphore,
+                                       Semaphore listenerSemaphore) {
+            this.streamIterator = streamIterator;
+            this.iteratorSemaphore = iteratorSemaphore;
+            this.listenerSemaphore = listenerSemaphore;
         }
-        onOpen.getRuntime().invokeMethodAsync(onOpen.getService(), onOpen.getFunctionName(), callback, properties,
-                computeMessageParams(onOpen, null, responseObserver));
-        callback.available.acquireUninterruptibly();
-
-        return new StreamObserver() {
-            @Override
-            public void onNext(Message value) {
-                ServiceResource onMessage = resourceMap.get(GrpcConstants.ON_MESSAGE_RESOURCE);
-
-                CallableUnitCallback callback = new StreamingCallableUnitCallBack(responseObserver, context);
-                onMessage.getRuntime().invokeMethodAsync(onMessage.getService(), onMessage.getFunctionName(),
-                        callback, properties, computeMessageParams(onMessage, value, responseObserver));
-            }
 
-            @Override
-            public void onError(Message error) {
-                ServiceResource onError = resourceMap.get(GrpcConstants.ON_ERROR_RESOURCE);
-                onErrorInvoke(onError, responseObserver, error, context);
+        @Override
+        public void onNext(Message value) {
+            try {
+                listenerSemaphore.acquire();
```

Comment:
If we can implement it using `Java Queue`, we can eliminate semaphore usage, right?

Reply:
Once we get a message, error, or eos, we add it to a queue, and we dequeue if in `next()`. This will also preserve the order